docs/proposals/20210614-podunavailablebudget.md

```diff
@@ -0,0 +1,148 @@
+---
+title: Pod Unavailable Budget
+authors:
+  - "@zmberg"
+reviewers:
+  - "@Fei-Guo"
+  - "@furykerry"
+  - "@FillZpp"
+creation-date: 2021-06-14
+last-updated: 2021-06-14
+status: implementable
+---
+
+# Pod Unavailable Budget Feature
+
+- In voluntary disruption scenarios, PodUnavailableBudget can achieve the effect of preventing application disruption or SLA degradation, which greatly improves the high availability of application services.
+
+## Table of Contents
+
+A table of contents is helpful for quickly jumping to sections of a proposal and for highlighting
+any additional information provided beyond the standard proposal template.
+[Tools for generating](https://github.com/ekalinin/github-markdown-toc) a table of contents from markdown are available.
+
+- [Title](#title)
+  - [Table of Contents](#table-of-contents)
+  - [Motivation](#motivation)
+  - [Proposal](#proposal)
+    - [API Definition](#api-definition)
+    - [Implement](#implement)
+    - [Program Comparison](#program-comparison)
+  - [User Cases](#user-cases)
+    - [A Sample Case](#a-sample-case)
+  - [Implementation History](#implementation-history)
+
+## Motivation
+
+In kubernetes, powerful workloads (e.g., deployment, statefulset, etc.) are often used to define and manage applications. It not only manages the number of pods, but also provides powerful rolling capabilities such as MaxUnavailable, Partition, etc. However, workload can only help the application reach the desired number of pods as fast as possible, but does not guarantee that the number of available pods will remain at the desired value at any given moment. In the following voluntary disruption scenarios, there are still business disruption and SLA degradation situations:
+1. When the owner deletes the workload by mistake, all the associated pods will be deleted
+2. When upgrading the application version, all pods are restarted at the same time due to the wrong configuration of the rolling upgrade policy, for example MaxUnavailable=0
```

Comment:
even if the user correctly set maxUnavailable in the workload, there is still cases where availability is hampered. For example when workload rolling upgrade and voluntary eviction of pods are both running. The key reason we use PUB is maxUnavailable in workload or PDB cannot protect simultaneous modification of pods. So I think we should hight light such user cases, and do not focus on the un-intentional workload deletion or ill-configured workload.